certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,214 @@
+    require threshold > 0 && to_mathint(threshold) <= guardianStorageContract.entries[safeContract].count + 1;
+
+    // Safe contract should be the sender of the transaction.
+    require e.msg.sender == safeContract;
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert !isReverted &&
+        currentContract.isGuardian(safeContract, guardian) &&
+        currentGuardiansCount + 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies the possibilites in which the addition of a new guardian can revert.
+rule addGuardianRevertPossibilities(env e, address guardian, uint256 threshold) {
+    bool isGuardian = currentContract.isGuardian(safeContract, guardian);
+
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert isReverted =>
+        isGuardian ||
+        e.msg.sender != safeContract ||
+        e.msg.value != 0 ||
+        guardian == 0 ||
+        guardian == SENTINEL() ||
+        guardian == safeContract ||
+        safeContract.isOwner(guardian) ||
+        threshold == 0 ||
+        to_mathint(threshold) > guardianStorageContract.entries[safeContract].count + 1 ||
+        guardianStorageContract.entries[safeContract].count == max_uint256 ||
+        !safeContract.isModuleEnabled(currentContract);
+}
+
+// This integrity rule verifies that if the revokeGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the guardian is revoked from the guardian list
+// - the linked list integrity remains,
+// - and no other account (guardian or not) is affected.
+rule revokeGuardiansWorksAsExpected(env e, address guardian, address prevGuardian, uint256 threshold, address otherAccount) {
+    requireGuardiansLinkedListIntegrity(guardian);
+
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+
+    currentContract.revokeGuardianWithThreshold(e, safeContract, prevGuardian, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert !currentContract.isGuardian(safeContract, guardian);
+    assert guardianStorageContract.entries[safeContract].guardians[prevGuardian] == nextGuardian;
+    assert guardian != otherAccount => otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount > 0 => currentGuardiansCount - 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
```

Comment:
Is this needed?
```suggestion
    assert currentGuardiansCount - 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
```

Reply:
`currentContract.revokeGuardianWithThreshold` should always revert if `currentGuardiansCount` is 0 no?

Reply:
Yes, that is correct. Updated 👍🏾